sandbox/Example.Core.SubscribeRaw/Program.cs

```diff
@@ -1,25 +1,82 @@
+using System.Diagnostics;
 using System.Text;
+using System.Threading.Channels;
 using Microsoft.Extensions.Logging;
 using NATS.Client.Core;
+using OpenTelemetry;
+using OpenTelemetry.Exporter;
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+
+using var trace = StartTracing();
+
+using var listener = new ActivityListener
+{
+    ActivityStarted = o => Console.WriteLine($"Started {o.OperationName} {o.Id}"),
+    ActivityStopped = o => Console.WriteLine($"Stopped {o.OperationName} {o.Id}"),
+    ShouldListenTo = o => o.Name == Telemetry.ActivitySourceName,
+};
+
+ActivitySource.AddActivityListener(listener);
 
 var subject = "foo.*";
 var options = NatsOpts.Default with { LoggerFactory = new MinimumConsoleLoggerFactory(LogLevel.Error) };
 
 Print("[CON] Connecting...\n");
+using var t = Telemetry.NatsActivities.StartActivity("Subscribing");
 
 await using var connection = new NatsConnection(options);
 
 Print($"[SUB] Subscribing to subject '{subject}'...\n");
 
 var sub = await connection.SubscribeAsync(subject);
 
-await foreach (var msg in sub.Msgs.ReadAllAsync())
+t.Dispose();
+
+await Task.WhenAll(Enumerable.Range(0, 3).Select(_ =>
+    ReadMessages(sub.Msgs, connection, default)));
+
+async Task ReadMessages(ChannelReader<NatsMsg> msgs, NatsConnection conn, CancellationToken ct)
 {
-    var data = Encoding.UTF8.GetString(msg.Data.ToArray());
-    Print($"[RCV] {msg.Subject}: {data}\n");
+    while (true)
+    {
+        try
+        {
+            var msg = await msgs.ReadAsync(ct);
+            using var _ = Activity.Current = msg.Activity;
```

Comment:
This isn't great, but since we have no way to know when `msg` goes out of scope, we have dispose/stop the activity via a `using` statement. Further, setting `Activity.Current` ensures context propagation on line 52.

Open to exposing this in a different way... but remember, `ChannelReader<NatsMsg>` can be consumed in parallel.

Reply:
I currently see two alternative options to this api here. @mtmk What are your thoughts on these two options?

## 1. Stop (i.e. "complete") the `Activity` when it is returned from `msgs.ReadAsync()`.
It is then on the user to instrument their application w/ some kind of `Processing` `Activity` (if they choose). I think we'd want to have a convenience method to propagate the context here. So user code might look something like:
```csharp
var msg = await msgs.ReadAsync(ct); // receive activity completes here

// ---- optional user-managed "process" activity  ----
// helper option A
using var processActivity = Activity.Current = msg.StartChildActivity("my_processing_activity");

// helper option B - user brings their own ActivitySource
using var processActivity = Activity.Current = MyActivitySource.StartActivity(
    "my_processing_activity",
    ActivityKind.Internal,
    parentContext: msg.GetActivityContext());
// ---------------------------------------------------


// ...do stuff with msg
``` 
#### Span Timing:
```markdown
| 'receive_message' span |
                          | 'my_processing_activity' span |
```


## 2. Wrap the message in a `scope` type which would be disposed by user code at end of processing.
This can be a `struct` to avoid an allocation. It might look something like this:
```csharp
using var msgScope = await msgs.ReadAsync(ct);

// ---- optional user-managed "process" activity  ----
// helper option A
using var processActivity = Activity.Current = msgScope.StartChildActivity("my_processing_activity");

// helper option B - user brings their own ActivitySource
using var processActivity = Activity.Current = MyActivitySource.StartActivity(
    "my_processing_activity",
    ActivityKind.Internal,
    parentContext: msgScope.GetActivityContext());
// -------------------------------------------------

var msg = msgScope.Msg;
// ...do stuff with msg
```
#### Span Timing:
```markdown
| ---- 'receive_message' span ------- |
     | 'my_processing_activity' span |
```

---
#### My Opinion
Personally I like the first option better because, in the case where the user doesn't want to deal w/ tracing, it doesn't get in their way. Further, this seems to better align w/ the OpenTelemetry [documentation and definitions](https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/messaging/#consumer). It aligns with the example given [here](https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/messaging/#apache-kafka-with-quarkus-or-spring-boot-example).

Reply:
I agree. First option seems better to me too.